gas-oracle/tokenprice/tokenprice.go

```diff
@@ -25,31 +63,82 @@ func NewClient(url string, frequency uint64) *Client {
 		return nil
 	})
 
+	uniswapQuoterClient, err := newUniswapClient(uniswapURL)
+	if err != nil {
+		return nil
+	}
+
 	return &Client{
-		client:    client,
-		frequency: time.Duration(frequency) * time.Second,
+		client:              client,
+		uniswapQuoterClient: uniswapQuoterClient,
+		frequency:           time.Duration(frequency) * time.Second,
+		tokenRatioMode:      TokenRatioMode(tokenRatioMode),
 	}
 }
 
-// Client is an HTTP based TokenPriceClient
-type Client struct {
-	client     *resty.Client
-	frequency  time.Duration
-	lastRatio  float64
-	lastUpdate time.Time
-}
+func (c *Client) PriceRatioWithMode() (float64, error) {
+	if time.Now().Sub(c.lastUpdate) < c.frequency {
+		return c.lastRatio, nil
+	}
 
-type TokenPrice struct {
-	Symbol string `json:"symbol"`
-	Price  string `json:"price"`
+	// Todo query token prices concurrent
+	var ratio, ethPrice float64
+	var errDex error
+	// get token price from dex
+	if ratio, ethPrice, errDex = c.getTokenPricesFromUniswap(); errDex != nil {
+		ratio = float64(DefaultTokenRatio)
+		ethPrice = float64(DefaultETHPrice)
+	}
+
+	// get token price from cex
+	// if both dex and cex return token prices correctly, token price from cex will be used
+	// if both failed, default value will be used
+	if ratioCex, ethPriceCex, errCex := c.priceRatio(); errCex != nil && errDex != nil {
+		ratio = float64(DefaultTokenRatio)
+		ethPrice = float64(DefaultETHPrice)
+	} else if errCex == nil {
+		ratio = ratioCex
+		ethPrice = ethPriceCex
+	}
+
+	switch c.tokenRatioMode {
+	case DefaultTokenRatioMode:
+		// use default eth/mnt price to set token ratio
+		ratio = float64(DefaultTokenRatio)
+	case OneDollarTokenRatioMode:
+		// supposing that mnt is 1 USD, so token_ratio is equals to eth_price
+		ratio = ethPrice
```

Comment:
Which mode is selected by default for use?  And how can the mode be dynamically changed during runtime?

Reply:
> Which mode is selected by default for use? And how can the mode be dynamically changed during runtime?

- Default mode is `RealTokenRatioMode`(`TokenRatioMode(0)`), which will use real eth_price/mnt_price to set token ratio. I will add a `default` here.
- Gasoracle currently does not support dynamic configuration replacement, and a reboot is required for the configuration to take effect. We will make further modifications in the future to support this feature.